okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -465,98 +464,104 @@ public void run() {
+                    sslSession == null ? SecurityLevel.NONE : SecurityLevel.PRIVACY_AND_INTEGRITY)
+                .build();
+          } catch (StatusException e) {
+            startGoAway(0, ErrorCode.INTERNAL_ERROR, e.getStatus());
+            return;
+          } catch (Exception e) {
+            onException(e);
+            return;
+          } finally {
+            clientFrameHandler = new ClientFrameHandler(variant.newReader(source, true));
+            executor.execute(clientFrameHandler);
           }
-          sock.setTcpNoDelay(true);
-          source = Okio.buffer(Okio.source(sock));
-          sink = Okio.buffer(Okio.sink(sock));
-          // The return value of OkHttpTlsUpgrader.upgrade is an SSLSocket that has this info
-          attributes = Attributes
-              .newBuilder()
-              .set(Grpc.TRANSPORT_ATTR_REMOTE_ADDR, sock.getRemoteSocketAddress())
-              .set(Grpc.TRANSPORT_ATTR_LOCAL_ADDR, sock.getLocalSocketAddress())
-              .set(Grpc.TRANSPORT_ATTR_SSL_SESSION, sslSession)
-              .set(GrpcAttributes.ATTR_SECURITY_LEVEL,
-                  sslSession == null ? SecurityLevel.NONE : SecurityLevel.PRIVACY_AND_INTEGRITY)
-              .build();
-        } catch (StatusException e) {
-          startGoAway(0, ErrorCode.INTERNAL_ERROR, e.getStatus());
-          return;
-        } catch (Exception e) {
-          onException(e);
-          return;
-        } finally {
-          clientFrameHandler = new ClientFrameHandler(variant.newReader(source, true));
-          executor.execute(clientFrameHandler);
-        }
 
-        FrameWriter rawFrameWriter;
-        synchronized (lock) {
-          socket = Preconditions.checkNotNull(sock, "socket");
-          maxConcurrentStreams = Integer.MAX_VALUE;
-          startPendingStreams();
-          if (sslSession != null) {
-            securityInfo = new InternalChannelz.Security(new InternalChannelz.Tls(sslSession));
+          synchronized (lock) {
+            socket = Preconditions.checkNotNull(sock, "socket");
+            maxConcurrentStreams = Integer.MAX_VALUE;
+            startPendingStreams();
+            if (sslSession != null) {
+              securityInfo = new InternalChannelz.Security(new InternalChannelz.Tls(sslSession));
+            }
           }
         }
+      });
 
-        rawFrameWriter = variant.newWriter(sink, true);
-        frameWriter.becomeConnected(rawFrameWriter, socket);
-
-        try {
-          // Do these with the raw FrameWriter, so that they will be done in this thread,
-          // and before any possible pending stream operations.
-          rawFrameWriter.connectionPreface();
-          Settings settings = new Settings();
-          rawFrameWriter.settings(settings);
-        } catch (Exception e) {
-          onException(e);
-          return;
-        }
-      }
-    });
+      // Do these with the transport lock, so that they will be done right after the start and
```

Comment:
This seems a bit weird of a comment, since any usage of `frameWriter` must be done with the transport lock. Maybe:

> Enqueue these immediately, so that they will be the first writes on the connection.